apps/nextjs/src/app/api/webhooks/clerk/route.ts

```diff
@@ -0,0 +1,95 @@
+import type { UserJSON } from "@clerk/backend";
+import type { WebhookEvent } from "@clerk/nextjs/server";
+import { posthogAiBetaServerClient } from "@oakai/core/src/analytics/posthogAiBetaServerClient";
+import { aiLogger } from "@oakai/logger";
+import * as Sentry from "@sentry/node";
+import { headers } from "next/headers";
+import { Webhook } from "svix";
+import invariant from "tiny-invariant";
+
+const WEBHOOK_SECRET = process.env.CLERK_WEBHOOK_SECRET;
+
+const log = aiLogger("webhooks");
+
+declare global {
+  interface UserPublicMetadata {
+    labs?: {
+      isDemoUser?: boolean;
+      isOnboarded?: boolean;
+      featureFlagGroup?: string | null;
+    };
+  }
+}
+
+function getPrimaryEmail(user: UserJSON): string {
+  const primaryEmail = user.email_addresses.find(
+    (email) => email.id === user.primary_email_address_id,
+  );
+  invariant(primaryEmail);
+  return primaryEmail.email_address;
+}
+
+async function syncUserToPosthog(user: UserJSON) {
+  const featureFlagGroup = user.public_metadata.labs?.featureFlagGroup ?? "";
+  posthogAiBetaServerClient.identify({
+    distinctId: getPrimaryEmail(user),
+    properties: { featureFlagGroup },
+  });
+  await posthogAiBetaServerClient.shutdown();
+  log.info("featureFlagGroup synced:", user.id, featureFlagGroup);
+}
+
+export async function POST(req: Request) {
+  try {
+    const headerPayload = headers();
+    const svixId = headerPayload.get("svix-id");
+    const svixTimestamp = headerPayload.get("svix-timestamp");
+    const svixSignature = headerPayload.get("svix-signature");
+
+    if (!svixId || !svixTimestamp || !svixSignature) {
+      return new Response("Error occured -- no svix headers", {
+        status: 400,
+      });
+    }
+
+    // Create a new Svix instance with your secret.
+    invariant(WEBHOOK_SECRET, "Missing CLERK_WEBHOOK_SECRET");
+    const wh = new Webhook(WEBHOOK_SECRET);
+
+    let evt: WebhookEvent;
+
+    try {
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      const payload = await req.json();
```

Comment:
Presumably we don't have a Zod schema available to us for this?

Reply:
I haven't found one provided by Clerk. You'll see a couple of lines below that they provide `as WebhookEvent` which is a really useful union. we'd just have to trust that they keep their end of the deal by not deviating from that type. It seems unlikely to me

We could validate just the fields we need with zod? Though I think if any were missing it would already throw an error

Reply:
Where do you think would be a good place to put in the schema validation without losing that more detailed type information? Maybe in `syncUserToPosthog`?

Reply:
I'm going to stick with it as-is, and if we find it needs more safety we can add it later. At the moment we're not using this for anything mission critical so it should be OK